verify: Support for verifying `numberio` in read-only verify phase

minwooim opened the pull request:
Add support for veryfing **latest** `numberio` in the verify header read from written offsets.
Previously, in the offline verification, read phase has not supported `numberio` compare since
it's not that simple to figure out the latest `numberio`s to the offsets.  This patch allows
verifying latest written verify header based on `numberio`.

Available test cases are the below:

- Sequential
```
fio --ioengine=io_uring_cmd --cmd_type=nvme --name=yeah --filename=/dev/ng0n1 --rw=write --bs=4k --size=32k --verify=crc32 --do_verify=0 --io_size=64k --verify_state_save=1

fio --ioengine=io_uring_cmd --thread=1 --cmd_type=nvme --name=yeah --filename=/dev/ng0n1 --rw=write --bs=4k --size=32k --verify=crc32 --do_verify=1 --io_size=64k --verify_state_load=1 --verify_only=1
```

- Random
```
fio --ioengine=io_uring_cmd --cmd_type=nvme --name=yeah --filename=/dev/ng0n1 --rw=randwrite --bs=4k --size=32k --verify=crc32 --do_verify=0 --io_size=64k --verify_state_save=1 --norandommap

fio --ioengine=io_uring_cmd --thread=1 --cmd_type=nvme --name=yeah --filename=/dev/ng0n1 --rw=randwrite --bs=4k --size=32k --verify=crc32 --do_verify=1 --io_size=64k --verify_state_load=1 --verify_only=1 --norandommap
```

minwooim commented:
@sitsofe , thw following patch made:

1) Dry run for wrapped around @io_u instances to log io pieces with the given numberio in the verify state file. (considered write only; no trim yet)
2) Put the io pieces to the rb-tree, not flist which does not consider overlaps.

This is not a clean-state patch, definitely need to revisit:

```diff
diff --git a/backend.c b/backend.c
index e09c210a..f205ea13 100644
--- a/backend.c
+++ b/backend.c
@@ -1686,10 +1686,14 @@ static int exec_string(struct thread_options *o, const char *string,
  */
 static uint64_t do_dry_run(struct thread_data *td)
 {
-       td_set_runstate(td, TD_RUNNING);
+       struct thread_io_list *s = td->vstate;
+       uint64_t target_numberio = s->numberio;
+
+       td_set_runstate(td, TD_DRY_RUNNING);

        while ((td->o.read_iolog_file && !flist_empty(&td->io_log_list)) ||
-               (!flist_empty(&td->trim_list)) || !io_complete_bytes_exceeded(td)) {
+               (!flist_empty(&td->trim_list)) || !io_complete_bytes_exceeded(td) ||
+               td->io_issues[DDIR_WRITE] < target_numberio) {
                struct io_u *io_u;
                int ret;

@@ -1723,6 +1727,8 @@ static uint64_t do_dry_run(struct thread_data *td)
                (void) ret;
        }

+       td_set_runstate(td, TD_RUNNING);
+
        return td->bytes_done[DDIR_WRITE] + td->bytes_done[DDIR_TRIM];
 }

diff --git a/fio.h b/fio.h
index 1fda07f8..efa57860 100644
--- a/fio.h
+++ b/fio.h
@@ -713,6 +713,7 @@ enum {
        TD_INITIALIZED,
        TD_RAMP,
        TD_SETTING_UP,
+       TD_DRY_RUNNING,
        TD_RUNNING,
        TD_PRE_READING,
        TD_VERIFYING,
diff --git a/io_u.c b/io_u.c
index 2d5986ec..91786ccb 100644
--- a/io_u.c
+++ b/io_u.c
@@ -390,6 +390,7 @@ static int get_next_seq_offset(struct thread_data *td, struct fio_file *f,
                }
         }

+retry:
        if (f->last_pos[ddir] < f->real_file_size) {
                uint64_t pos;

@@ -430,6 +431,10 @@ static int get_next_seq_offset(struct thread_data *td, struct fio_file *f,

                *offset = pos;
                return 0;
+       } else if (td->runstate == TD_DRY_RUNNING) {
+               f->last_pos[ddir] = f->file_offset;
+               loop_cache_invalidate(td, f);
+               goto retry;
        }

        return 1;
diff --git a/iolog.c b/iolog.c
index dcf6083c..27e393a4 100644
--- a/iolog.c
+++ b/iolog.c
@@ -306,7 +306,7 @@ void log_io_piece(struct thread_data *td, struct io_u *io_u)
         * the rb insert/lookup for handling. Sort writes if we have offset
         * modifier which can also create duplicate blocks.
         */
-       if (!fio_offset_overlap_risk(td)) {
+       if (!fio_offset_overlap_risk(td) && td->runstate != TD_DRY_RUNNING) {
                INIT_FLIST_HEAD(&ipo->list);
                flist_add_tail(&ipo->list, &td->io_hist_list);
                ipo->flags |= IP_F_ONLIST;
diff --git a/libfio.c b/libfio.c
index 57f3f858..7570d695 100644
--- a/libfio.c
+++ b/libfio.c
@@ -209,7 +209,7 @@ static const char *td_runstates[] = {

 const char *runstate_to_name(int runstate)
 {
-       compiletime_assert(TD_LAST == 12, "td runstate list");
+       compiletime_assert(TD_LAST == 13, "td runstate list");
        if (runstate >= 0 && runstate < TD_LAST)
                return td_runstates[runstate];
```

vincentkfu reviewed (commented):
The first change to `fio_offset_overlap_risk` looks good to me.
The second change probably is ok as well. However, I'm not comfortable letting it in because `get_next_seq_offset()` has to handle many different situations and we have very little testing for multiple files, holed IO, backwards IO, etc.

vincentkfu commented:
For now I think it's possible to accomplish the aim of this PR by explicitly setting `verify_write_sequence=1`. Do you agree?

minwooim commented on io_u.c:
```diff
@@ -383,7 +383,7 @@ static int get_next_seq_offset(struct thread_data *td, struct fio_file *f,
 	 * and invalidate the cache, if we need to.
 	 */
-	if (td_rw(td) && o->io_size > o->size) {
+	if (o->io_size > o->size) {
```
This change has been from d782b76fc9086dac25d674a9439352f9fa962db1 and it's trying not to wrap-around too early in the rw case.  read or write workloads will be checked in the next if-statement based on the `real_file_size`, so we can keep it as it was.

minwooim commented on fio.h:
```diff
@@ -826,7 +826,8 @@ extern void clear_inflight(struct thread_data *);
 	if (td->o.norandommap || td->o.softrandommap ||
-	    td->o.ddir_seq_add || (td->o.ddir_seq_nr > 1))
+	    td->o.ddir_seq_add || (td->o.ddir_seq_nr > 1) ||
+	    td->o.io_size > td->o.size)
```
Yes, `io_size` will be treated per-file, so multiple files are fine with this change to detect overlap risks.

minwooim committed:
fio.h: treat io_size > size as offset overlap risk

When td->o.io_size > td->o.size, sequential writes wrap around and
revisit the same offsets.  In this case, ``fio_offset_overlap_risk()``
must return true so that io_hist uses an rb-tree instaed of a plain
flist, allowing overallping io_pieces to be detected and replaced
correctly.

This check may produce false positives for multi-file jobs where
per-file wrap-around does not actually occur, but that is acceptable
for the simplicity and since the only cost is using rb-tree
unnecessarily.  A false negative, however, would silently corrupt the
verification history.

Signed-off-by: Minwoo Im <minwoo.im@samsung.com>

minwooim committed:
io_u: check wrap around f->last_pos if verify_only=1

--verify_only=1 ensures to follow the sequence of write phases of the
current jobfile.  Not only just for --rw=rw case, but --verify_only=1
case should also consider wrapping around the f->last_pos to 0 when it
reaches to the end of the f->io_size to make sure that the READ offsets
get to have latest `numberio` value in which overlapped WRITEs were
happening.

Offline verification for the overlapped sequential write offsets now can
be verified properly:

- Write Phase

./fio --name=test --ioengine=io_uring_cmd --filename=/dev/ng0n1 \
	--cmd_type=nvme --rw=write --bs=128k --size=1M --io_size=2M \
	--verify=crc32 --do_verify=0 --verify_state_save=1 --debug=io

io       254431 complete: io_u 0x79eb20000d80: off=0x0,len=0x20000,ddir=1,file=/dev/ng0n1
io       254431 complete: io_u 0x79eb20000d80: off=0x20000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254431 complete: io_u 0x79eb20000d80: off=0x40000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254431 complete: io_u 0x79eb20000d80: off=0x60000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254431 complete: io_u 0x79eb20000d80: off=0x80000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254431 complete: io_u 0x79eb20000d80: off=0xa0000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254431 complete: io_u 0x79eb20000d80: off=0xc0000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254431 complete: io_u 0x79eb20000d80: off=0xe0000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254431 complete: io_u 0x79eb20000d80: off=0x0,len=0x20000,ddir=1,file=/dev/ng0n1
io       254431 complete: io_u 0x79eb20000d80: off=0x20000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254431 complete: io_u 0x79eb20000d80: off=0x40000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254431 complete: io_u 0x79eb20000d80: off=0x60000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254431 complete: io_u 0x79eb20000d80: off=0x80000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254431 complete: io_u 0x79eb20000d80: off=0xa0000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254431 complete: io_u 0x79eb20000d80: off=0xc0000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254431 complete: io_u 0x79eb20000d80: off=0xe0000,len=0x20000,ddir=1,file=/dev/ng0n1

- Verify Phase

Dry run with WRITE offsets with overlap-aware, and do the ddir=0(READ)
phase with the latest `numberio` checks.

./fio --name=test --ioengine=io_uring_cmd --filename=/dev/ng0n1 \
	--cmd_type=nvme --rw=write --bs=128k --size=1M --io_size=2M \
	--verify=crc32 --do_verify=1 --verify_only --verify_state_load=1

io       254628 complete: io_u 0x5aa26dc2dac0: off=0x0,len=0x20000,ddir=1,file=/dev/ng0n1
io       254628 complete: io_u 0x5aa26dc2dac0: off=0x20000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254628 complete: io_u 0x5aa26dc2dac0: off=0x40000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254628 complete: io_u 0x5aa26dc2dac0: off=0x60000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254628 complete: io_u 0x5aa26dc2dac0: off=0x80000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254628 complete: io_u 0x5aa26dc2dac0: off=0xa0000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254628 complete: io_u 0x5aa26dc2dac0: off=0xc0000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254628 complete: io_u 0x5aa26dc2dac0: off=0xe0000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254628 iolog: overlap 0/131072, 0/131072
io       254628 complete: io_u 0x5aa26dc2dac0: off=0x0,len=0x20000,ddir=1,file=/dev/ng0n1
io       254628 iolog: overlap 131072/131072, 131072/131072
io       254628 complete: io_u 0x5aa26dc2dac0: off=0x20000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254628 iolog: overlap 262144/131072, 262144/131072
io       254628 complete: io_u 0x5aa26dc2dac0: off=0x40000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254628 iolog: overlap 393216/131072, 393216/131072
io       254628 complete: io_u 0x5aa26dc2dac0: off=0x60000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254628 iolog: overlap 524288/131072, 524288/131072
io       254628 complete: io_u 0x5aa26dc2dac0: off=0x80000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254628 iolog: overlap 655360/131072, 655360/131072
io       254628 complete: io_u 0x5aa26dc2dac0: off=0xa0000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254628 iolog: overlap 786432/131072, 786432/131072
io       254628 complete: io_u 0x5aa26dc2dac0: off=0xc0000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254628 iolog: overlap 917504/131072, 917504/131072
io       254628 complete: io_u 0x5aa26dc2dac0: off=0xe0000,len=0x20000,ddir=1,file=/dev/ng0n1
io       254628 complete: io_u 0x5aa26dc2dac0: off=0x0,len=0x20000,ddir=0,file=/dev/ng0n1
io       254628 complete: io_u 0x5aa26dc2dac0: off=0x20000,len=0x20000,ddir=0,file=/dev/ng0n1
io       254628 complete: io_u 0x5aa26dc2dac0: off=0x40000,len=0x20000,ddir=0,file=/dev/ng0n1
io       254628 complete: io_u 0x5aa26dc2dac0: off=0x60000,len=0x20000,ddir=0,file=/dev/ng0n1
io       254628 complete: io_u 0x5aa26dc2dac0: off=0x80000,len=0x20000,ddir=0,file=/dev/ng0n1
io       254628 complete: io_u 0x5aa26dc2dac0: off=0xa0000,len=0x20000,ddir=0,file=/dev/ng0n1
io       254628 complete: io_u 0x5aa26dc2dac0: off=0xc0000,len=0x20000,ddir=0,file=/dev/ng0n1
io       254628 complete: io_u 0x5aa26dc2dac0: off=0xe0000,len=0x20000,ddir=0,file=/dev/ng0n1

Also, for write-only sequential jobs, `total_bytes` was capped at size
regardless of `io_size`, causing the write phase to stop after one pass
over the file even when `io_size` > `size`.  Unlike rw mode where reads
and writes both consume `bytes_issued`, write-only jobs only count
writes, so there is no risk of `io_size` being consumed by reads.  Allow
`io_size` to control how much to write for this case.

This allows online verification when `io_size` > `size` with write-only
sequential jobs.

	fio \
	--name=online \
	--ioengine=io_uring_cmd --filename=/dev/ng0n1 \
	--cmd_type=nvme \
	--rw=write --bs=128k --size=1M --io_size=2M \
	--verify=crc32 --do_verify=1 --debug=io,verify

Before this patch:

- Writes didn't happen twice (no overlap) even though `io_size` >
  `size` since reads consumed the `bytes_issued`.

io       238438 complete: io_u 0x74a23c000d80: off=0x0,len=0x20000,ddir=1,file=/dev/ng0n1
io       238438 complete: io_u 0x74a23c000d80: off=0x20000,len=0x20000,ddir=1,file=/dev/ng0n1
io       238438 complete: io_u 0x74a23c000d80: off=0x40000,len=0x20000,ddir=1,file=/dev/ng0n1
io       238438 complete: io_u 0x74a23c000d80: off=0x60000,len=0x20000,ddir=1,file=/dev/ng0n1
io       238438 complete: io_u 0x74a23c000d80: off=0x80000,len=0x20000,ddir=1,file=/dev/ng0n1
io       238438 complete: io_u 0x74a23c000d80: off=0xa0000,len=0x20000,ddir=1,file=/dev/ng0n1
io       238438 complete: io_u 0x74a23c000d80: off=0xc0000,len=0x20000,ddir=1,file=/dev/ng0n1
io       238438 complete: io_u 0x74a23c000d80: off=0xe0000,len=0x20000,ddir=1,file=/dev/ng0n1
io       238438 complete: io_u 0x74a23c000d80: off=0x0,len=0x20000,ddir=0,file=/dev/ng0n1
io       238438 complete: io_u 0x74a23c000d80: off=0x20000,len=0x20000,ddir=0,file=/dev/ng0n1
io       238438 complete: io_u 0x74a23c000d80: off=0x40000,len=0x20000,ddir=0,file=/dev/ng0n1
io       238438 complete: io_u 0x74a23c000d80: off=0x60000,len=0x20000,ddir=0,file=/dev/ng0n1
io       238438 complete: io_u 0x74a23c000d80: off=0x80000,len=0x20000,ddir=0,file=/dev/ng0n1
io       238438 complete: io_u 0x74a23c000d80: off=0xa0000,len=0x20000,ddir=0,file=/dev/ng0n1
io       238438 complete: io_u 0x74a23c000d80: off=0xc0000,len=0x20000,ddir=0,file=/dev/ng0n1
io       238438 complete: io_u 0x74a23c000d80: off=0xe0000,len=0x20000,ddir=0,file=/dev/ng0n1

After this patch:

- Writes are overlapped, but verify once with the latest `numberio`
  verification.

io       237335 complete: io_u 0x71c1f4000d80: off=0x0,len=0x20000,ddir=1,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0x20000,len=0x20000,ddir=1,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0x40000,len=0x20000,ddir=1,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0x60000,len=0x20000,ddir=1,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0x80000,len=0x20000,ddir=1,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0xa0000,len=0x20000,ddir=1,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0xc0000,len=0x20000,ddir=1,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0xe0000,len=0x20000,ddir=1,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0x0,len=0x20000,ddir=1,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0x20000,len=0x20000,ddir=1,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0x40000,len=0x20000,ddir=1,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0x60000,len=0x20000,ddir=1,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0x80000,len=0x20000,ddir=1,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0xa0000,len=0x20000,ddir=1,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0xc0000,len=0x20000,ddir=1,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0xe0000,len=0x20000,ddir=1,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0x0,len=0x20000,ddir=0,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0x20000,len=0x20000,ddir=0,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0x40000,len=0x20000,ddir=0,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0x60000,len=0x20000,ddir=0,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0x80000,len=0x20000,ddir=0,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0xa0000,len=0x20000,ddir=0,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0xc0000,len=0x20000,ddir=0,file=/dev/ng0n1
io       237335 complete: io_u 0x71c1f4000d80: off=0xe0000,len=0x20000,ddir=0,file=/dev/ng0n1

Signed-off-by: Minwoo Im <minwoo.im@samsung.com>

minwooim commented:
@vincentkfu , PR has been revisit, please kindly review.

vincentkfu reviewed (dismissed):
The first patch seems fine.

vincentkfu reviewed (commented):
Regarding the second patch, I think some more needs to be done if you want to ensure that fio wraps around reliably once it reaches `size`. Your example appears to be a niche case with io_uring_cmd where the actual device size is larger than `size`.

With psync and a regular file, if size and filesize two are equal, fio will not wrap around.

```
$ rm test.0.0 && ./fio --name=test --rw=write --bs=512k --size=1M --io_size=2M --verify=crc32c --do_verify=0 --filesize=1M --debug=io
io: set debug option io
test: (g=0): rw=write, bs=(R) 512KiB-512KiB, (W) 512KiB-512KiB, (T) 512KiB-512KiB, ioengine=psync, iodepth=1
fio-3.41-142-g698f
Starting 1 process
io       6102  declare unneeded cache test.0.0: 0/1048576
io       6102  fill: io_u 0x557718ae8f00: off=0x0,len=0x80000,ddir=1,file=test.0.0
io       6102  prep: io_u 0x557718ae8f00: off=0x0,len=0x80000,ddir=1,file=test.0.0
io       6102  queue: io_u 0x557718ae8f00: off=0x0,len=0x80000,ddir=1,file=test.0.0
io       6102  complete: io_u 0x557718ae8f00: off=0x0,len=0x80000,ddir=1,file=test.0.0
io       6102  fill: io_u 0x557718ae8f00: off=0x80000,len=0x80000,ddir=1,file=test.0.0
io       6102  prep: io_u 0x557718ae8f00: off=0x80000,len=0x80000,ddir=1,file=test.0.0
io       6102  queue: io_u 0x557718ae8f00: off=0x80000,len=0x80000,ddir=1,file=test.0.0
io       6102  complete: io_u 0x557718ae8f00: off=0x80000,len=0x80000,ddir=1,file=test.0.0
io       6102  get_next_offset: offset 1048576 >= io_size 1048576
io       6102  io_u 0x557718ae8f00, failed getting offset
io       6102  io_u 0x557718ae8f00, setting file failed
io       6102  get_io_u failed
io       6102  io_u 0x557718ae8f00, setting file failed
io       6102  get_io_u failed
io       6102  close ioengine psync
io       6102  free ioengine psync

test: (groupid=0, jobs=1): err= 0: pid=6102: Thu Mar 19 19:44:26 2026
  write: IOPS=2000, BW=1000MiB/s (1049MB/s)(1024KiB/1msec)
    clat (usec): min=81, max=101, avg=91.66, stdev=13.84
     lat (usec): min=202, max=214, avg=208.25, stdev= 8.25
    clat percentiles (usec):
     |  1.00th=[   82],  5.00th=[   82], 10.00th=[   82], 20.00th=[   82],
     | 30.00th=[   82], 40.00th=[   82], 50.00th=[   82], 60.00th=[  102],
     | 70.00th=[  102], 80.00th=[  102], 90.00th=[  102], 95.00th=[  102],
     | 99.00th=[  102], 99.50th=[  102], 99.90th=[  102], 99.95th=[  102],
     | 99.99th=[  102]
  lat (usec)   : 100=50.00%, 250=50.00%
  cpu          : usr=0.00%, sys=0.00%, ctx=0, majf=0, minf=20
  IO depths    : 1=100.0%, 2=0.0%, 4=0.0%, 8=0.0%, 16=0.0%, 32=0.0%, >=64=0.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     issued rwts: total=0,2,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0.00ns, window=0.00ns, percentile=100.00%, depth=1

Run status group 0 (all jobs):
  WRITE: bw=1000MiB/s (1049MB/s), 1000MiB/s-1000MiB/s (1049MB/s-1049MB/s), io=1024KiB (1049kB), run=1-1msec

Disk stats (read/write):
  vda: ios=0/0, sectors=0/0, merge=0/0, ticks=0/0, in_queue=0, util=0.00%
```

With psync and a regular file and `filesize` greater than `size`, fio appears to simply ignore `size`.

```
$ rm test.0.0 && ./fio --name=test --rw=write --bs=512k --size=1M --io_size=2M --verify=crc32c --do_verify=0 --filesize=2M --debug=io
fio: set debug option io
test: (g=0): rw=write, bs=(R) 512KiB-512KiB, (W) 512KiB-512KiB, (T) 512KiB-512KiB, ioengine=psync, iodepth=1
fio-3.41-142-g698f
Starting 1 process
test: Laying out IO file (1 file / 2MiB)
io       6151  declare unneeded cache test.0.0: 0/2097152
io       6153  declare unneeded cache test.0.0: 0/2097152
io       6153  fill: io_u 0x5624d8364f00: off=0x0,len=0x80000,ddir=1,file=test.0.0
io       6153  prep: io_u 0x5624d8364f00: off=0x0,len=0x80000,ddir=1,file=test.0.0
io       6153  queue: io_u 0x5624d8364f00: off=0x0,len=0x80000,ddir=1,file=test.0.0
io       6153  complete: io_u 0x5624d8364f00: off=0x0,len=0x80000,ddir=1,file=test.0.0
io       6153  fill: io_u 0x5624d8364f00: off=0x80000,len=0x80000,ddir=1,file=test.0.0
io       6153  prep: io_u 0x5624d8364f00: off=0x80000,len=0x80000,ddir=1,file=test.0.0
io       6153  queue: io_u 0x5624d8364f00: off=0x80000,len=0x80000,ddir=1,file=test.0.0
io       6153  complete: io_u 0x5624d8364f00: off=0x80000,len=0x80000,ddir=1,file=test.0.0
io       6153  fill: io_u 0x5624d8364f00: off=0x100000,len=0x80000,ddir=1,file=test.0.0
io       6153  prep: io_u 0x5624d8364f00: off=0x100000,len=0x80000,ddir=1,file=test.0.0
io       6153  queue: io_u 0x5624d8364f00: off=0x100000,len=0x80000,ddir=1,file=test.0.0
io       6153  complete: io_u 0x5624d8364f00: off=0x100000,len=0x80000,ddir=1,file=test.0.0
io       6153  fill: io_u 0x5624d8364f00: off=0x180000,len=0x80000,ddir=1,file=test.0.0
io       6153  prep: io_u 0x5624d8364f00: off=0x180000,len=0x80000,ddir=1,file=test.0.0
io       6153  queue: io_u 0x5624d8364f00: off=0x180000,len=0x80000,ddir=1,file=test.0.0
io       6153  complete: io_u 0x5624d8364f00: off=0x180000,len=0x80000,ddir=1,file=test.0.0
io       6153  close ioengine psync
io       6153  free ioengine psync

test: (groupid=0, jobs=1): err= 0: pid=6153: Thu Mar 19 19:48:12 2026
  write: IOPS=2000, BW=1000MiB/s (1049MB/s)(2048KiB/2msec)
    clat (usec): min=98, max=114, avg=104.67, stdev= 6.89
     lat (usec): min=247, max=272, avg=260.72, stdev=12.97
    clat percentiles (usec):
     |  1.00th=[   98],  5.00th=[   98], 10.00th=[   98], 20.00th=[   98],
     | 30.00th=[  101], 40.00th=[  101], 50.00th=[  101], 60.00th=[  105],
     | 70.00th=[  105], 80.00th=[  115], 90.00th=[  115], 95.00th=[  115],
     | 99.00th=[  115], 99.50th=[  115], 99.90th=[  115], 99.95th=[  115],
     | 99.99th=[  115]
  lat (usec)   : 100=25.00%, 250=75.00%
  cpu          : usr=0.00%, sys=0.00%, ctx=0, majf=0, minf=17
  IO depths    : 1=100.0%, 2=0.0%, 4=0.0%, 8=0.0%, 16=0.0%, 32=0.0%, >=64=0.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     issued rwts: total=0,4,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0.00ns, window=0.00ns, percentile=100.00%, depth=1

Run status group 0 (all jobs):
  WRITE: bw=1000MiB/s (1049MB/s), 1000MiB/s-1000MiB/s (1049MB/s-1049MB/s), io=2048KiB (2097kB), run=2-2msec

Disk stats (read/write):
  vda: ios=0/0, sectors=0/0, merge=0/0, ticks=0/0, in_queue=0, util=0.00%
```

vincentkfu commented on io_u.c:
```diff
@@ -383,11 +383,11 @@ static int get_next_seq_offset(struct thread_data *td, struct fio_file *f,
+	 * should get back to 0 offset to verify overlapped offsets.
 	 */
-	if ((td_rw(td) || o->verify_only) && o->io_size > o->size) {
+	if ((td_rw(td) || (o->verify != VERIFY_NONE && o->do_verify)) &&
```
This removes the change from the previous patch. Why have the previous patch at all if you remove the change here?

minwooim replied:
This was intended.  This change is for online-verification.  `o->verify_only=1` is valid in offline verification in verify phase, but this condition is no sufficient to cover the online verification.  Otherwise, online verification will not go back to 0 offset.

vincentkfu commented:
I think we should start fresh with a new PR because there is no clear connection between the commit messages for these patches and the original motivation for this PR.

What is the purpose of these patches? Originally this PR sought to enable numberio verification in `verify_only` workloads but none of the commit message examples have `verify_write_sequence=1` as far as I can tell.

Instead these patches try to change how fio's behavior depends on `size` and `io_size`. I believe the purpose of these patcfhes is still `numberio` verification. Can you start a new PR and write a cover letter explaining how these patches connect to `numberio` verification and include example configurations in the commit messages with `verify_write_sequence=1`?

Also, it would help to add test cases to ensure that the use cases you are interested in do not become broken.

minwooim commented:
> I think we should start fresh with a new PR because there is no clear connection between the commit messages for these patches and the original motivation for this PR.
> 
> What is the purpose of these patches? Originally this PR sought to enable numberio verification in `verify_only` workloads but none of the commit message examples have `verify_write_sequence=1` as far as I can tell.
>
> Instead these patches try to change how fio's behavior depends on `size` and `io_size`. I believe the purpose of these patcfhes is still `numberio` verification. Can you start a new PR and write a cover letter explaining how these patches connect to `numberio` verification and include example configurations in the commit messages with `verify_write_sequence=1`?
> 
> Also, it would help to add test cases to ensure that the use cases you are interested in do not become broken.

Let me have a new PR for this with considering `filesize` > `size` along with test cases.  Thanks for the review!

minwooim committed:
libfio: reset nr_done_files when io_size > size

reset_io_counters() clears td->nr_done_files so that keep_running()
does not return false prematurely because fio_files_done() sees all
files already marked done.

The existing conditions (time_based, loops > 1, do_verify) cover the
cases where the job is expected to restart file iteration.  A job with
io_size > size also requires restarting: the sequential write pointer
wraps around and visits every offset a second time, so the "file done"
bit must be cleared at the start of each outer iteration.

Without this fix, the first pass sets each fio_file's done flag, and
keep_running() exits the outer loop early instead of writing the
second pass -- the overlap that the io_size > size feature is meant to
produce never happens.

Signed-off-by: Minwoo Im <minwoo.im@samsung.com>

minwooim committed:
t/numberio_overlap: add overlap write verification tests

Add a Python test suite testing fio's numberio overlap verification for
the three fixes that preceded this commit:

  - fio.h: treat io_size > size as offset overlap risk
  - io_u.c: check wrap around f->last_pos if verify_only=1
  - backend.c: use io_size as limit for seq write-only
  - libfio.c: reset nr_done_files when io_size > size

The test creates situations where io_size > size so that every offset
is written more than once.  fio_offset_overlap_risk() must return true
to activate the rb-tree io_hist backend, which retains only the latest
io_piece per offset.  The verify phase then reads each block exactly
once and checks that numberio on disk matches the latest write.

18 test cases, grouped by rw mode / verify style / I/O engine:

  TC  1- 2  rw=write,     offline verify,  io_uring + psync
  TC  3- 4  rw=write,     online  verify,  io_uring + psync
  TC  5- 6  rw=randwrite, offline verify,  io_uring + psync  (norandommap)
  TC  7- 8  rw=randwrite, online  verify,  io_uring + psync  (norandommap)
  TC  9-10  rw=rw,        offline verify,  io_uring + psync
  TC 11-12  rw=rw,        online  verify,  io_uring + psync
  TC 13-14  rw=randrw,    offline verify,  io_uring + psync  (norandommap)
  TC 15-16  rw=randrw,    online  verify,  io_uring + psync  (norandommap)
  TC 17-18  rw=write,     offline verify,  io_uring + psync  (filesize < size)

Offline verify (OfflineOverlapVerifyTest) runs two separate fio
invocations: a write phase (verify_state_save=1) followed by a verify
phase (verify_only=1, verify_state_load=1).  The verify phase performs
a dry-run write pass to rebuild io_hist, then reads each block and
checks numberio via verify_write_sequence=1.

Online verify (OnlineOverlapVerifyTest) runs a single fio invocation
with do_verify=1, which writes and verifies in the same job.

The filesize < size cases (TC 17-18) pre-truncate the file smaller than
size= before running fio, verifying that fio_offset_overlap_risk()
activates the rb-tree even before the first I/O because real_file_size
< io_size at setup time.

Signed-off-by: Minwoo Im <minwoo.im@samsung.com>
